rajawali/src/main/java/c/org/rajawali3d/sceneview/SceneView.java

```diff
@@ -0,0 +1,227 @@
+        } else {
+            onEngineCleared();
+        }
+    }
+
+    @Nullable
+    @Override
+    public RenderModel getRenderModel() {
+        return scene;
+    }
+
+    @Override
+    public void setViewport(Rect viewport) {
+        this.viewport.set(viewport);
+    }
+
+    @Override
+    public Rect getViewport() {
+        return viewport;
+    }
+
+    @Override
+    public int getViewportWidth() {
+        return viewport.width();
+    }
+
+    @Override
+    public int getViewportHeight() {
+        return viewport.height();
+    }
+
+    @Override
+    public void isDisbled(boolean disabled) {
+        this.isDisabled = disabled;
+    }
+
+    @Override
+    public boolean isDisabled() {
+        return false;
+    }
+
+
+    @GLThread
+    @Override
+    public void onFrameStart(final double deltaTime)
+            throws InterruptedException {
+        // TODO is a scene lock actually needed here?  This may all be independent of the Scene...
+        // TODO is a sceneView lock needed here?
+        //currentlyHeldReadLock = scene.acquireReadLock();
+        try {
+            // TODO frame tasks for changes to viewport and/or render passes
+            // Execute frame tasks
+            //performFrameTasks();
+
+            // TODO view-specific frame callbacks could be useful, e.g. enabling viewport animations
+        } finally {
+            if (currentlyHeldReadLock != null) {
+                currentlyHeldReadLock.unlock();
+            }
+        }
+    }
+
+    @GLThread
+    @Override
+    public void onRenderView() throws InterruptedException {
+        // TODO Lock should probably be on the whole Scene, not just its Graph
+        //currentlyHeldReadLock = scene.acquireReadLock();
+        try {
+            internalRender();
+        } finally {
+            if (currentlyHeldReadLock != null) {
+                currentlyHeldReadLock.unlock();
+            }
+        }
+    }
+
+    @Override
+    public void onFrameEnd(double deltaTime) throws InterruptedException {
+        // TODO Anything?
+    }
+
+    @RequiresReadLock
+    @GLThread
+    protected void internalRender() throws IllegalStateException {
+
+        // TODO Probably need to mimic the nextCameraLock pattern for viewport changes?
+        //updateProjectionMatrix(width, height);
+        //GLES20.glViewport(0, 0, width, height);
+
+        synchronized (nextCameraLock) {
+            // Check if we need to switch the camera, and if so, do it.
+            if (nextCamera != null) {
+                switchCamera(nextCamera);
+                nextCamera = null;
+            }
+        }
+
+        // Prepare the camera matrices
+        final Matrix4 viewMatrix = currentCamera.getViewMatrix();
+        final Matrix4 projectionMatrix = currentCamera.getProjectionMatrix();
+
+        if (projectionMatrix == null) {
+            throw new IllegalStateException("Cannot render while current camera has a null projection matrix.");
```

Comment:
This exception was wrong in my initial code and I have since changed it to an error log and break out. We will want to update this to be in line with that as well.